src/routes/add-article.ts

```diff
@@ -1,33 +1,51 @@
+import clownface from 'clownface';
 import createHttpError from 'http-errors';
 import { constants } from 'http2';
-import jsonld from 'jsonld';
-import { JsonLdArray } from 'jsonld/jsonld-spec';
 import { Next } from 'koa';
-import { schema } from 'rdf-namespaces';
+import { Quad } from 'rdf-js';
 import { termToString } from 'rdf-string';
 import uniqueString from 'unique-string';
 import url from 'url';
 import { AppContext, AppMiddleware } from '../app';
 import NotAnArticle from '../errors/not-an-article';
+import { rdf, schema } from '../namespaces';
 import Routes from './index';
 
 export default (): AppMiddleware => (
   async ({
-    articles, dataFactory: { blankNode }, request, response, router,
+    articles, dataFactory: { blankNode, quad }, request, response, router,
   }: AppContext, next: Next): Promise<void> => {
-    const [article] = await jsonld.expand(request.body) as JsonLdArray;
-    if ('@id' in article) {
-      throw new createHttpError.Forbidden(`Article IDs must not be set ('${article['@id']}' was given)`);
+    const id = clownface({ dataset: request.dataset }).has(rdf.type, schema.Article).term;
+
+    if (!id) {
+      throw new createHttpError.BadRequest(`No ${termToString(schema.Article)} found`);
+    }
+
+    if (id.termType !== 'BlankNode') {
+      throw new createHttpError.BadRequest(`Article must have a blank node identifier (${termToString(id)} given)`);
     }
-    if (!(schema.name in article) || article[schema.name].length === 0) {
-      throw new createHttpError.BadRequest(`Article must have at least one ${schema.name}`);
+
+    if (request.dataset.match(id, schema('name')).size === 0) {
+      throw new createHttpError.BadRequest(`Article must have at least one ${termToString(schema('name'))}`);
     }
 
     const newId = blankNode(uniqueString());
-    article['@id'] = termToString(newId);
+
+    [...request.dataset].forEach((originalQuad: Quad): void => {
+      let newQuad: Quad;
+      if (originalQuad.subject.equals(id)) {
+        newQuad = quad(newId, originalQuad.predicate, originalQuad.object, originalQuad.graph);
+      } else if (originalQuad.object.equals(id)) {
+        newQuad = quad(originalQuad.subject, originalQuad.predicate, newId, originalQuad.graph);
+      } else {
+        return;
```

Comment:
ignores silently all data in the graph that is not associated to the article id in the role of subject or object?

Reply:
Leaves them as is. Other option is to create a new dataset, so there's always an add.

Reply:
Good as-is following must forward.